src/Utils.ts

```diff
@@ -1645,28 +1658,42 @@ class DummyInputManagerCore {
         }
 
         if (this._isOutside) {
-            const elementParent = element.parentElement;
+            const elementParent = dom.getParentNode(element);
 
             if (elementParent) {
-                const nextSibling = element.nextElementSibling;
+                const nextSibling = dom.getNextSibling(element);
 
                 if (nextSibling !== lastDummyInput) {
-                    elementParent.insertBefore(lastDummyInput, nextSibling);
+                    dom.insertBefore(
+                        elementParent,
+                        lastDummyInput,
+                        nextSibling
+                    );
                 }
 
-                if (element.previousElementSibling !== firstDummyInput) {
-                    elementParent.insertBefore(firstDummyInput, element);
+                if (
+                    dom.getPreviousElementSibling(element) !== firstDummyInput
+                ) {
+                    dom.insertBefore(elementParent, firstDummyInput, element);
                 }
             }
         } else {
-            if (element.lastElementChild !== lastDummyInput) {
-                element.appendChild(lastDummyInput);
+            if (dom.getLastElementChild(element) !== lastDummyInput) {
+                dom.appendChild(element, lastDummyInput);
             }
 
-            const firstElementChild = element.firstElementChild;
+            const firstElementChild = dom.getFirstElementChild(element);
 
-            if (firstElementChild && firstElementChild !== firstDummyInput) {
-                element.insertBefore(firstDummyInput, firstElementChild);
+            if (
+                firstElementChild &&
+                firstElementChild !== firstDummyInput &&
+                firstElementChild.parentNode
+            ) {
+                dom.insertBefore(
+                    firstElementChild.parentNode,
```

Comment:
This too?

Reply:
This is a special case. If we need to insert dummy inputs around top-level element of a shadow root, they need to be inserted inside the shadow, not inside the host. Adding anything else to the shadow host (other than attached shadow) technically works, the elements appear in the DOM, but those elements are ignored by the browser, they are not rendered and they don't receive focus events.